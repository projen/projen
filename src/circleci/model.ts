// @see https://circleci.com/docs/2.0/configuration-reference

/**
 * Used to represent objects with a dynamic key.
 * Dynamic are marked with `<>` in configuration reference like '<workflow_name>' and '<job_name>'
 * This field is required for objects with dynamic keys.
 */
interface INamed {
  /** name of dynamic key **/
  readonly identifier: string;
}

/**
 * Used for orchestrating all jobs. Each workflow consists of the workflow name as a key and a map as a value.
 * A name should be unique within the current config.yml.
 * The top-level keys for the Workflows configuration are version and jobs.
 * @see https://circleci.com/docs/2.0/configuration-reference/#workflows
 */
export interface Workflow extends INamed {
  readonly jobs?: WorkflowJob[];
  readonly triggers?: Triggers[];
  /** when is too dynamic to be casted to interfaces. Check Docu as reference
   * @see https://circleci.com/docs/2.0/configuration-reference/#logic-statement-examples
   */
  readonly when?: any;
}

/**
 * A job may have a type of approval indicating it must be manually approved before downstream jobs may proceed
 * @see https://circleci.com/docs/2.0/configuration-reference/#type
 */
export enum JobType {
  APPROVAL = "approval",
}

/**
 * The resource_class feature allows configuring CPU and RAM resources for each job.
 * Different resource classes are available for different executors, as described in the tables below.
 * @see https://circleci.com/docs/2.0/configuration-reference/#resourceclass
 */
export enum ResourceClass {
  SMALL = "small",
  MEDIUM = "medium",
  MEDIUM_PLUS = "medium+",
  LARGE_X = "xlarge",
  LARGE_2X = "2xlarge",
  LARGE_2X_PLUS = "2xlarge+",
}

/**
 * Specify when to enable or disable the step.
 * @see https://circleci.com/docs/2.0/configuration-reference/#steps
 */
export enum JobWhen {
  ALWAYS = "always",
  ON_SUCCESS = "on_success",
  ON_FAIL = "on_fail",
}

/**
 * Pipeline parameter types
 * @see https://circleci.com/docs/2.0/reusing-config#parameter-syntax
 */
export enum PipelineParameterType {
  STRING = "string",
  BOOLEAN = "boolean",
  INTEGER = "integer",
  ENUM = "enum",
}

/**
 * A Workflow is comprised of one or more uniquely named jobs. Jobs are specified in the jobs map,
 * see Sample 2.0 config.yml for two examples of a job map.
 * The name of the job is the key in the map, and the value is a map describing the job.
 * @see https://circleci.com/docs/2.0/configuration-reference/#jobs
 */
export interface Job extends INamed {}

/**
 * A Job is part of Workflow. A Job can be created with {@link Job} or it can be provided by the orb
 * @see https://circleci.com/docs/2.0/configuration-reference/#jobs-in-workflow
 */
export interface WorkflowJob extends INamed {
  /** A list of jobs that must succeed for the job to start. */
  readonly requires?: string[];
  /** A replacement for the job name. Useful when calling a job multiple times */
  readonly name?: string;
  /** The name of the context(s). The initial default name is org-global. Each context name must be unique. */
  readonly context?: string[];
  /** A job may have a type of approval indicating it must be manually approved before downstream jobs may proceed. */
  readonly type?: JobType;
  /** Job Filters can have the key branches or tags */
  readonly filters?: Filter;
  readonly matrix?: Matrix;
  /** Parameters passed to job when referencing a job from orb */
  readonly orbParameters?: Record<string, string | number | boolean>;
}

/**
 * Parameters are declared by name under a job, command, or executor.
 * @see https://circleci.com/docs/2.0/reusing-config#using-the-parameters-declaration
 */
export interface PipelineParameter {
  /** Used to generate documentation for your orb. */
  readonly description?: string;
  /** The parameter type, required. */
  readonly type: PipelineParameterType;
  /** The default value for the parameter. If not present, the parameter is implied to be required. */
  readonly default?: string | number | boolean;
}

/**
 * The matrix stanza allows you to run a parameterized job multiple times with different arguments.
 * @see https://circleci.com/docs/2.0/configuration-reference/#matrix-requires-version-21
 */
export interface Matrix {
  /** A map of parameter names to every value the job should be called with */
  readonly parameters?: Record<string, string[] | number[]>;
  /** An alias for the matrix, usable from another job’s requires stanza. Defaults to the name of the job being executed */
  readonly alias?: string;
}

/**
 * Specifies which triggers will cause this workflow to be executed.
 * Default behavior is to trigger the workflow when pushing to a branch.
 * @see https://circleci.com/docs/2.0/configuration-reference/#triggers
 */
export interface Triggers {
  readonly schedule?: Schedule;
}

/**
 *  A workflow may have a schedule indicating it runs at a certain time.
 *  @see https://circleci.com/docs/2.0/configuration-reference/#schedule
 */
export interface Schedule {
  /** The cron key is defined using POSIX crontab syntax. */
  readonly cron?: string;
  readonly filters: Filter;
}

/**
 * The branches key controls whether the current branch should have a schedule trigger created for it,
 * where current branch is the branch containing the config.yml file with the trigger stanza.
 * That is, a push on the main branch will only schedule a workflow for the main branch.
 *
 * Branches can have the keys only and ignore which either map to a single string naming a branch.
 * You may also use regular expressions to match against branches by enclosing them with /’s, or map to a list of such strings.
 * Regular expressions must match the entire string.
 *
 * Any branches that match only will run the job.
 * Any branches that match ignore will not run the job.
 * If neither only nor ignore are specified then all branches will run the job.
 * If both only and ignore are specified the only is considered before ignore.
 * @see https://circleci.com/docs/2.0/configuration-reference/#filters
 */
export interface Filter {
  readonly branches?: FilterConfig;
  readonly tags?: FilterConfig;
}

/**
 * set an inclusive or exclusive filter
 * @see https://circleci.com/docs/2.0/configuration-reference/#filters
 */
export interface FilterConfig {
  /** Either a single branch specifier, or a list of branch specifiers */
  readonly only?: string[];
  /** Either a single branch specifier, or a list of branch specifiers */
  readonly ignore?: string[];
}

/**
 * Each job consists of the job’s name as a key and a map as a value. A name should be case insensitive unique within a current jobs list.
 * @see https://circleci.com/docs/2.0/configuration-reference/#job_name
 */
export interface Job {
  readonly docker?: Docker[];
  readonly machine?: Machine;
  readonly macos?: Macos;
  /** Shell to use for execution command in all steps. Can be overridden by shell in each step */
  readonly shell?: string;
  /** Parameters for making a job explicitly configurable in a workflow. */
  readonly parameters?: Record<string, PipelineParameter>;
  /** no type support here, for syntax {@see https://circleci.com/docs/2.0/configuration-reference/#steps} */
  readonly steps?: any[];
  /** In which directory to run the steps. Will be interpreted as an absolute path. Default: `~/project` */
  readonly workingDirectory?: string;
  /** Number of parallel instances of this job to run (default: 1) */
  readonly parallelism?: number;
  /** A map of environment variable names and values. */
  readonly environment?: Record<string, string | number | boolean>;
  /** {@link ResourceClass} */
  readonly resourceClass?: ResourceClass | string;
  /** Enables jobs to go through a set of well-defined IP address ranges */
  readonly circleciIpRanges?: boolean;
}

/**
 * Options for docker executor
 * @see https://circleci.com/docs/2.0/configuration-reference/#docker
 */
export interface Docker {
  /** The name of a custom docker image to use */
  readonly image: string;
  /** The name the container is reachable by. By default, container services are accessible through localhost */
  readonly name?: string;
  /** The command used as executable when launching the container */
  readonly entrypoint?: string[];
  /** The command used as pid 1 (or args for entrypoint) when launching the container */
  readonly command?: string[];
  /** Which user to run commands as within the Docker container */
  readonly user?: string;
  /** A map of environment variable names and values */
  readonly environment?: Record<string, string | number | boolean>;
  /** Authentication for registries using standard docker login credentials */
  readonly auth?: Record<string, string>;
  /** Authentication for AWS Elastic Container Registry (ECR) */
  readonly awsAuth?: Record<string, string>;
}

export interface Machine {
  /**
   * The VM image to use.
   * @see https://circleci.com/docs/2.0/configuration-reference/#available-machine-images
   */
  readonly image: string;
  /** enable docker layer caching
   * @see https://circleci.com/docs/2.0/configuration-reference/#available-machine-images
   */
  readonly dockerLayerCaching?: string;
}

/**
 * CircleCI supports running jobs on macOS, to allow you to build, test,
 * and deploy apps for macOS, iOS, tvOS and watchOS. To run a job in a macOS virtual machine,
 * you must add the macos key to the top-level configuration for the job and specify
 * the version of Xcode you would like to use.
 * @see https://circleci.com/docs/2.0/configuration-reference/#macos
 */
export interface Macos {
  /** The version of Xcode that is installed on the virtual machine */
  readonly xcode: string;
}

/**
 * Execution steps for Job
 * @see https://circleci.com/docs/2.0/configuration-reference/#steps
 */
export interface StepRun {
  readonly run?: Run;
}

/**
 * Used for invoking all command-line programs, taking either a map of configuration values,
 * or, when called in its short-form, a string that will be used as both the command and name.
 * Run commands are executed using non-login shells by default,
 * so you must explicitly source any dotfiles as part of the command.
 *
 * Not used because type incompatible types in steps array
 *
 * @see https://circleci.com/docs/2.0/configuration-reference/#run
 */
export interface Run {
  /** Command to run via the shell */
  readonly command: string;
  /** Title of the step to be shown in the CircleCI UI (default: full command) */
  readonly name?: string;
  /** Shell to use for execution command */
  readonly shell?: string;
  /** Additional environmental variables, locally scoped to command */
  readonly environment?: string;
  /** Whether this step should run in the background (default: false) */
  readonly background?: string;
  /** In which directory to run this step. Will be interpreted relative to the working_directory of the job). (default: .) */
  readonly workingDirectory?: string;
  /** Elapsed time the command can run without output such as “20m”, “1.25h”, “5s”. The default is 10 minutes */
  readonly noOutputTimeout?: string;
  /** Specify when to enable or disable the step. */
  readonly when?: JobWhen | string;
}
