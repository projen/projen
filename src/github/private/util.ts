export function secretToString(secretName: string): string {
  return `\${{ secrets.${secretName} }}`;
}

export function context(value: string) {
  return `\${{ ${value} }}`;
}

// Checks if part of the file path is hidden
export function isHiddenPath(path: string) {
  return /(^|\/)\.[^\/\.]/g.test(path);
}

// Helper to assert a path is not hidden
export function ensureNotHiddenPath(value: string, name: string) {
  if (isHiddenPath(value)) {
    throw Error(`${name} cannot be a hidden path, got: ${value}`);
  }
}
