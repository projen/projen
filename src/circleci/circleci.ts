import { WorkflowJob, Workflow, Job } from "./model";
import { Component } from "../component";
import { Project } from "../project";
import { snakeCaseKeys } from "../util";
import { YamlFile } from "../yaml";

/**
 * Options for class {@link Circleci}
 *
 * @see https://circleci.com/docs/2.0/configuration-reference/
 */
export interface CircleCiProps {
  /**
   * Contains a map of CirclCi Orbs
   * ```json
   * orbs: {
   *  node: "circleci/node@5.0.1"
   *  slack: "circleci/slack@4.8.3"
   * }
   * ```
   */
  readonly orbs?: Record<string, string>;
  /**
   * pipeline version
   *
   * @default 2.1
   * @see https://circleci.com/docs/2.0/configuration-reference/#version
   */
  readonly version?: number;
  /**
   * List of Workflows of pipeline, e.g.
   * ```json
   * workflows: {
   *   {
   *     identifier: "build",
   *       jobs: [{
   *          identifier: "node/install",
   *          context: ["npm"],
   *       }]
   *   }
   * }
   * ```
   *
   * @see https://circleci.com/docs/2.0/configuration-reference/#workflows
   */
  readonly workflows?: Workflow[];
  /**
   * List of Jobs to create unique steps per pipeline, e.g.
   * ```json
   * jobs: [{
   *  identifier: "compile",
   *  docker: { image: "golang:alpine" }
   *  steps: ["checkout", run: {command: "go build ."}]
   * }]
   * ```
   *
   * @see https://circleci.com/docs/2.0/configuration-reference/#jobs
   */
  readonly jobs?: Job[];
  /**
   * The setup field enables you to conditionally trigger configurations from outside
   * the primary .circleci parent directory, update pipeline parameters, or generate customized configurations.
   *
   * @see https://circleci.com/docs/2.0/configuration-reference/#setup
   */
  readonly setup?: boolean;
}

/**
 * Circleci Class to manage `.circleci/config.yml`.
 * Check projen's docs for more information.
 *
 * @see https://circleci.com/docs/2.0/configuration-reference/
 */
export class Circleci extends Component {
  /**
   * The yaml file for the Circleci pipeline
   * */
  public readonly file: YamlFile;
  /**
   * internal copy of options to share options between functions
   * @private
   */
  private options: CircleCiProps;
  /**
   * internal map of orbs
   * @private
   */
  private readonly orbs: Record<string, string>;
  /**
   * internal list of workflows
   * @private
   */
  private workflows: Workflow[];
  /**
   * internal list of jobs
   * @private
   */
  private readonly jobs: Job[];

  constructor(project: Project, options: CircleCiProps = {}) {
    super(project);
    this.options = options;
    this.orbs = options.orbs ?? {};
    this.workflows = options.workflows ?? [];
    this.jobs = options.jobs ?? [];
    this.file = new YamlFile(project, ".circleci/config.yml", {
      obj: () => this.renderCircleCi(),
    });
  }

  /**
   * function to prepare the rendering of the yaml file.
   * Objects with dynamic keys like workflows and jobs required the field `identifier` to be set.
   * Those object will be reduced by `identifier` field before rendering
   * @private
   */
  private renderCircleCi() {
    // render dynamic keys for workflow
    const workflowRecords: Record<string, any> = {};
    for (const workflow of this.workflows) {
      const { identifier, ...reduced } = workflow;
      reduced.jobs = this.renderJobs(workflow.jobs);
      workflowRecords[identifier] = this.snakeCase(reduced);
    }

    // render dynamic keys for jobs
    const jobRecords: Record<string, any> = {};
    for (const job of this.jobs) {
      const { identifier, ...reduced } = job;
      jobRecords[identifier] = this.snakeCase(reduced);
    }

    return {
      version: this.options.version || 2.1,
      setup: this.options.setup,
      orbs: this.orbs,
      jobs: Object.keys(jobRecords).length > 0 ? jobRecords : undefined,
      workflows: workflowRecords,
    };
  }

  /**
   * add new workflow to existing pipeline
   * @param workflow
   */
  public addWorkflow(workflow: Workflow) {
    this.workflows = [...this.workflows, workflow];
  }

  /**
   * reduce objects with `identifier` field for WorkflowJobs.
   * A workflow job can contain `orbParameter` which are passed to orbs.
   * This map is directly added as `Record<string,any>` to job.
   * So we gonna add those after the default field of WorkflowJob.
   * @see https://circleci.com/developer/orbs/orb/circleci/node#usage-install_nodejs
   * @param jobs
   */
  private renderJobs = (jobs: WorkflowJob[] = []): any => {
    let result: any = [];
    for (const job of jobs ?? []) {
      const { identifier, orbParameters, ...reduced } = job;
      if (isObjectContainingFieldExactly(job, "identifier")) {
        result = [...result, identifier];
      } else {
        result = [
          ...result,
          { [identifier]: { ...reduced, ...orbParameters } },
        ];
      }
    }
    return result;
  };

  /**
   * Add a Circleci Orb to pipeline. Will throw error if the orb already exists
   * @param name
   * @param orb
   */
  public addOrb(name: string, orb: string) {
    if (this.orbs[name] !== undefined) {
      throw new Error(`Circleci Config already contains an orb named ${name}.`);
    }
    this.orbs[name] = orb;
  }

  /**
   * Snake case for listed keys. There are too many exceptions needed to do it recursive without a whitelist.
   * This list needs to be updated once we add field with snake case like `aws_auth`.
   * @param input
   */
  private snakeCase = (input: any): any => {
    const snakeCaseKeyWords = [
      "awsAuth",
      "workingDirectory",
      "resourceClass",
      "dockerLayerCaching",
      "noOutputTimeout",
      "circleciIpRanges",
    ];
    return snakeCaseKeys(input, true, snakeCaseKeyWords);
  };
}

export const isObjectContainingFieldExactly = (
  obj: any,
  field: string
): boolean => {
  return Object.keys(obj).length == 1 && Object.keys(obj).includes(field);
};
