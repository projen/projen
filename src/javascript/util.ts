import { existsSync, readFileSync } from "fs";
import { basename, dirname, extname, join, sep, resolve, posix } from "path";
import * as semver from "semver";
import { NodePackage } from "./node-package";
import { Project } from "../project";
import { findUp } from "../util";

/**
 * Basic interface for `package.json`.
 */
interface PackageManifest {
  [key: string]: any;

  /**
   * Package name.
   */
  name: string;
  /**
   * Package version.
   */
  version: string;
}

export function renderBundleName(entrypoint: string) {
  const parts = join(entrypoint).split(sep);
  if (parts[0] === "src") {
    parts.shift(); // just remove 'src' if its the first element for ergonomics
  }

  const p = parts.join(posix.sep);
  const dir = dirname(p);
  const base = basename(p, extname(p));
  return posix.join(dir, base);
}

/**
 * Regex for AWS CodeArtifact registry
 */
export const codeArtifactRegex =
  /^https:\/\/(?<registry>(?<domain>[^\.]+)-(?<accountId>\d{12})\.d\.codeartifact\.(?<region>[^\.]+).*\.amazonaws\.com\/.*\/(?<repository>[^\/]+)\/)/;

/**
 * gets AWS details from the Code Artifact registry URL
 * throws exception if not matching expected pattern
 * @param registryUrl Code Artifact registry URL
 * @returns object containing the (domain, accountId, region, repository)
 */
export function extractCodeArtifactDetails(registryUrl: string) {
  const match = registryUrl.match(codeArtifactRegex);
  if (match?.groups) {
    const { domain, accountId, region, repository, registry } = match.groups;
    return { domain, accountId, region, repository, registry };
  }
  throw new Error("Could not get CodeArtifact details from npm Registry");
}

export function minVersion(version: string): string | undefined {
  if (semver.validRange(version)) {
    return semver.minVersion(version)?.version;
  } else {
    return version;
  }
}

/**
 * Attempt to resolve location of the given `moduleId`.
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveModule(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  try {
    return require.resolve(moduleId, options);
  } catch {
    return undefined;
  }
}

/**
 * Attempt to resolve a module's manifest (package.json) path via `require.resolve` lookup.
 *
 * @remarks
 * If the target package has `exports` that differ from the default
 * (i.e, it defines the `exports` field in its manifest) and does not
 * explicitly include an entry for `package.json`, this strategy will fail.
 * See {@link tryResolveManifestPathFromDefaultExport} as an alternative.
 *
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveModuleManifestPath(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  // cannot just `require('dependency/package.json')` here because
  // `options.paths` may not overlap with this node proc's resolution paths.
  const manifestId = `${moduleId}/package.json`;
  return tryResolveModule(manifestId, options);
}

/**
 * Attempt to resolve a module's manifest (package.json) path by looking for the nearest
 * `package.json` file that is an ancestor to the module's default export location.
 *
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveManifestPathFromDefaultExport(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  const defaultExportPath = tryResolveModule(moduleId, options);
  if (!defaultExportPath) {
    return undefined;
  }
  const moduleDir = findUp("package.json", defaultExportPath);
  if (!moduleDir) {
    return undefined;
  }
  return join(moduleDir, "package.json");
}

/**
 * Attempt to resolve a module's manifest (package.json) path by checking for its existence under `node_modules` relative to `basePath`.
 *
 * @remarks
 * This strategy can be helpful in the scenario that a module defines
 * custom exports without `package.json` and no default export (i.e, some type definition packages).
 *
 * @param moduleId Module ID to lookup.
 * @param basePath Root path to search from.
 */
export function tryResolveManifestPathFromPath(
  moduleId: string,
  basePath: string
) {
  const base = basePath.includes("node_modules")
    ? basePath
    : join(basePath, "node_modules");
  const filePath = resolve(base, ...moduleId.split("/"), "package.json");
  if (existsSync(filePath)) {
    return filePath;
  }
  return undefined;
}

/**
 * Attempt to resolve a module's manifest (package.json) path by searching for it in the optionally provided paths array
 * as well as the current node processes' default resolution paths.
 * @param moduleId Module ID to search for.
 * @param options Search options.
 */
export function tryResolveManifestPathFromSearch(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  const searchPaths = [
    ...(options?.paths ?? []),
    ...(require.resolve.paths(moduleId) ?? []),
  ];
  for (const path of searchPaths) {
    const result = tryResolveManifestPathFromPath(moduleId, path);
    // early return on first result.
    if (result) {
      return result;
    }
  }
  return undefined;
}

/**
 * Attempt to resolve a module's manifest (package.json) using multiple strategies.
 * @param moduleId Module to resolve manifest path for.
 * @param options Resolution options.
 */
export function tryResolveModuleManifest(
  moduleId: string,
  options?: { paths: string[] }
): PackageManifest | undefined {
  const strategies = [
    tryResolveModuleManifestPath,
    tryResolveManifestPathFromDefaultExport,
    tryResolveManifestPathFromSearch,
  ];
  for (const strategy of strategies) {
    const result = strategy(moduleId, options);
    // early return on first result.
    if (result) {
      try {
        const manifest = JSON.parse(
          readFileSync(result, "utf8")
        ) as PackageManifest;
        // verify name matches target module.
        if (manifest.name === moduleId) {
          return manifest;
        }
      } catch {
        // continue to next strategy.
      }
    }
  }
  return undefined;
}

/**
 * Attempt to resolve the installed version of a given dependency.
 * @param dependencyName Name of dependency.
 * @param options Optional options passed through to `require.resolve`.
 */
export function tryResolveDependencyVersion(
  dependencyName: string,
  options?: { paths: string[] }
): string | undefined {
  const manifest = tryResolveModuleManifest(dependencyName, options);
  if (!manifest) {
    return undefined;
  }
  return manifest?.version;
}

/**
 * Whether the given dependency version is installed
 *
 * This can be used to test for the presence of certain versions of devDependencies,
 * and do something dependency-specific in certain Components. For example, test for
 * a version of Jest and generate different configs based on the Jest version.
 *
 * NOTE: The implementation of this function currently is currently
 * approximate: to do it correctly, we would need a separate implementation
 * for every package manager, to query its installed version (either that, or we
 * would code to query `package-lock.json`, `yarn.lock`, etc...).
 *
 * Instead, we will look at `package.json`, and assume that the versions
 * picked by the package manager match ~that. This will work well enough for
 * major version checks, but may fail for point versions.
 *
 * What we SHOULD do is: `actualVersion ∈ checkRange`.
 *
 * What we do instead is a slightly more sophisticated version of
 * `requestedRange ∩ checkRange != ∅`. This will always give a correct result if
 * `requestedRange ⊆ checkRange`, but may give false positives when `checkRange
 * ⊆ requestedRange`.
 *
 * May return `undefined` if the question cannot be answered. These include the
 * following cases:
 *
 *   - The dependency is requested via local file dependencies (`file://...`)
 *   - The dependency uses an other type of URL, such as a GitHub URL
 *   - The dependency is not found in the `package.json`, such as when
 *     bootstrapping from an external projen package, and the `package.json`
 *     file only has that one external package as a dependency
 *
 * Otherwise it will return `true` if the installed version is probably in the
 * requested range, and `false` if it is probably not.
 *
 * This API may eventually be added to the public projen API, but only after
 * we implement exact version checking.
 *
 * @param dependencyName The name of the dependency
 * @param checkRange A particular version, or range of versions.
 */
export function hasDependencyVersion(
  project: Project,
  dependencyName: string,
  checkRange: string
): boolean | undefined {
  const file = NodePackage.of(project)?.file;
  if (!file) {
    return undefined;
  }

  if (!existsSync(file.absolutePath)) {
    return undefined;
  }

  const pj = JSON.parse(readFileSync(file.absolutePath, "utf-8"));

  // Technicaly, we should be intersecting all ranges to come up with the most narrow dependency
  // range, but `semver` doesn't allow doing that and we don't want to add a dependency on `semver-intersect`.
  //
  // Let's take the first dependency declaration we find, and assume that people
  // set up their `package.json` correctly.
  let requestedRange: string | undefined;
  for (const key of ["dependencies", "devDependencies", "peerDependencies"]) {
    const deps = pj[key] ?? {};
    let requestedVersion = deps[dependencyName];
    if (requestedVersion) {
      // If this is not a valid range, it could be 'file:dep.tgz', or a GitHub URL. No way to know what
      // version we're getting, bail out.
      if (!semver.validRange(requestedVersion)) {
        return undefined;
      }
      requestedRange = requestedVersion;
      break;
    }
  }

  // If the dependency is not found in the `package.json`, we can't answer the question (yet).
  // This means that the dependency hasn't been added yet, which means we know (upstream) what we're going to request,
  // or we're going to ask for '*' and we'll get the latest version.
  if (!requestedRange) {
    return undefined;
  }

  return installedVersionProbablyMatches(requestedRange, checkRange);
}

/**
 * Whether the given requestedRange *probably* leads to the installation of a version that matches checkRange
 *
 * We assume that NPM always installs the most recent version of a package that
 * is allowed by the requestedRange.
 */
export function installedVersionProbablyMatches(
  requestedRange: string,
  checkRange: string
): boolean {
  const options = {
    includePrerelease: true,
    loose: true,
  };

  // No questions asked: always true
  if (semver.subset(requestedRange, checkRange, options)) {
    return true;
  }

  // Also no questions asked: always false
  if (!semver.intersects(requestedRange, checkRange, options)) {
    return false;
  }

  // Now we're in tricky territory. We intersect, but aren't a full subset.
  // We are in one of the following 2 situations, which we will tie-break by
  // assuming NPM will install the most recent matching version in 'requested'.
  //
  // requested  | check    | result
  // -----------|----------|-----------
  //   >= 2     |  >= 3    | probably true (chance of FP)
  //   <= 2     |  <= 1    | probably false (change of FN)
  //
  // `semver` doesn't make it easy to distinguish these cases (we can't request
  // the `maxVersion` that satisfies a range). Instead what we do is
  // get the `minVersion` of each range, and if they compare equal we assume
  // we're in the bottom case with `<=` checks, and return `false`.

  return !semver.eq(
    semver.minVersion(requestedRange, options) ?? "1.2.3",
    semver.minVersion(checkRange, options) ?? "1.2.3"
  );
}
