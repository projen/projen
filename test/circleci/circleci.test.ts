import * as YAML from "yaml";
import { WorkflowJob } from "../../lib/circleci";
import {
  Circleci,
  CircleCiProps,
  isObjectContainingFieldExactly,
  PipelineParameterType,
} from "../../src/circleci";
// @ts-ignore
import { synthSnapshot, TestProject } from "../util";

test("full spec of api should be provided", () => {
  // GIVEN
  const p = new TestProject();
  const options: CircleCiProps = {
    orbs: {
      hello: "world:2.0",
    },
    jobs: [
      {
        identifier: "custom-job-1",
        docker: [
          {
            image: "golang:alpine",
            environment: {
              GO111MODULE: true,
            },
          },
        ],
        workingDirectory: ".",
        machine: {
          image: "node:alpine",
        },
        steps: [
          "checkout",
          {
            run: {
              command: ["echo $HELLOW_WORLD", "echo this is multiline"].join(
                "\n"
              ),
            },
          },
        ],
        parameters: {
          job1Param: {
            description: "job1Param description",
            type: PipelineParameterType.STRING,
            default: "job1Param default",
          },
        },
        circleciIpRanges: true,
      },
    ],
    workflows: [
      {
        identifier: "workflow1",
        when: {
          or: [
            { equal: ["main", "<< pipeline.git.branch >>"] },
            { equal: ["staging", "<< pipeline.git.branch >>"] },
          ],
        },
        triggers: [
          {
            schedule: {
              cron: "0 0 * * *",
              filters: {
                branches: {
                  only: ["main", "beta"],
                },
              },
            },
          },
        ],
        jobs: [
          {
            identifier: "job1",
            name: "renamedJob2",
            orbParameters: {
              "install-yarn": true,
              node_version: 16.13,
            },
            matrix: {
              parameters: {
                version: [0.1, 0.2, 0.3],
                platform: ["macos", "windows", "linux"],
              },
            },
          },
          {
            identifier: "checkout",
          },
        ],
      },
    ],
  };
  new Circleci(p, options);
  const snapshot = synthSnapshot(p);
  const circleci = snapshot[".circleci/config.yml"];
  const yaml = YAML.parse(circleci);

  expect(circleci).toMatchSnapshot();
  expect(circleci).toContain("renamedJob2");
  expect(circleci).toContain("0 0 * * *");
  expect(circleci).toContain("windows");
  const wJob = yaml.workflows.workflow1.jobs[0].job1;
  // testing orb parameters and snake case should be ignored here
  expect(wJob).toEqual(
    expect.objectContaining({
      "install-yarn": true,
      node_version: 16.13,
    })
  );

  const customJob = yaml.jobs["custom-job-1"];
  expect(customJob.docker[0].image).toEqual("golang:alpine");
  expect(customJob.machine.image).toEqual("node:alpine");
  expect(customJob.circleci_ip_ranges).toEqual(true);
  expect(customJob.steps).toHaveLength(2);
  expect(customJob.steps).toContain("checkout");
  expect(customJob).toEqual(
    // test snake case
    expect.objectContaining({ working_directory: "." })
  );
  expect(customJob).toEqual(
    // test parameters
    expect.objectContaining({
      parameters: {
        job1Param: {
          description: "job1Param description",
          type: "string",
          default: "job1Param default",
        },
      },
    })
  );
});

test("test type conversion for workflow jobs with identifier only", () => {
  const job1: WorkflowJob = {
    identifier: "checkout",
  };
  const job2: WorkflowJob = {
    identifier: "checkout",
    name: "hello-world",
  };
  expect(isObjectContainingFieldExactly(job1, "identifier")).toEqual(true);
  expect(isObjectContainingFieldExactly(job2, "identifier")).toEqual(false);
});

test("additional workflow can be added", () => {
  // GIVEN
  const p = new TestProject();
  const options: CircleCiProps = {
    orbs: {
      hello: "world:2.0",
    },
    workflows: [
      {
        identifier: "workflow1",
        jobs: [
          {
            identifier: "job1",
            context: ["context1"],
          },
          {
            identifier: "job2",
            name: "customName2",
          },
        ],
      },
    ],
  };
  const circle = new Circleci(p, options);
  circle.addWorkflow({
    identifier: "workflow3",
    jobs: [{ identifier: "job3", context: ["context3"] }],
  });
  const snapshot = synthSnapshot(p);
  const circleci = snapshot[".circleci/config.yml"];
  expect(circleci).toMatchSnapshot();
  expect(circleci).toContain("customName2");
  expect(circleci).toContain("context3");
});

test("orb with the same id can not be added", () => {
  // GIVEN
  const p = new TestProject();
  const options: CircleCiProps = {
    orbs: {
      hello: "world:2.0",
    },
    workflows: [
      {
        identifier: "workflow1",
        jobs: [
          {
            identifier: "job1",
            name: "renamedJob2",
          },
        ],
      },
    ],
  };
  const circle = new Circleci(p, options);
  expect(() => circle.addOrb("hello", "world:3.0")).toThrowError(
    "Circleci Config already contains an orb named hello."
  );

  const snapshot = synthSnapshot(p);
  const circleci = snapshot[".circleci/config.yml"];
  expect(circleci).toMatchSnapshot();
  expect(circleci).toContain("workflow1");
  expect(circleci).toContain("world:2.0");
  expect(circleci).not.toContain("world:3.0");
});
