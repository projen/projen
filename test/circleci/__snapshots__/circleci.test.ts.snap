// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`additional workflow can be added 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

version: 2.1
orbs:
  hello: world:2.0
workflows:
  workflow1:
    jobs:
      - job1:
          context:
            - context1
      - job2:
          name: customName2
  workflow3:
    jobs:
      - job3:
          context:
            - context3
"
`;

exports[`full spec of api should be provided 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

version: 2.1
orbs:
  hello: world:2.0
jobs:
  custom-job-1:
    docker:
      - image: golang:alpine
        environment:
          GO111MODULE: true
    working_directory: .
    machine:
      image: node:alpine
    steps:
      - checkout
      - run:
          command: |-
            echo $HELLOW_WORLD
            echo this is multiline
    parameters:
      job1Param:
        description: job1Param description
        type: string
        default: job1Param default
    circleci_ip_ranges: true
workflows:
  workflow1:
    when:
      or:
        - equal:
            - main
            - << pipeline.git.branch >>
        - equal:
            - staging
            - << pipeline.git.branch >>
    triggers:
      - schedule:
          cron: 0 0 * * *
          filters:
            branches:
              only:
                - main
                - beta
    jobs:
      - job1:
          name: renamedJob2
          matrix:
            parameters:
              version:
                - 0.1
                - 0.2
                - 0.3
              platform:
                - macos
                - windows
                - linux
          install-yarn: true
          node_version: 16.13
      - checkout
"
`;

exports[`orb with the same id can not be added 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

version: 2.1
orbs:
  hello: world:2.0
workflows:
  workflow1:
    jobs:
      - job1:
          name: renamedJob2
"
`;
